CMakeLists.txt

```diff
@@ -115,7 +115,7 @@ else()
   set(WITH_ACCELERATED_CPU false)
 endif()
 
-if(WITH_SSCP_COMPILER) # sanitize input
+if(WITH_SSCP_COMPILER OR WITH_METAL_BACKEND) # sanitize input
```

Comment:
Do we really need to special case metal here? `WITH_SSCP_COMPILER` is automatically enabled if LLVM is sufficiently new and it's not explicitly disabled. `WITH_XXX_BACKEND` refers to runtime backends, not necessarily compiler infrastructure.

The other backends work as follows:
* `WITH_SSCP_COMPILER` is activated by default for sufficiently new LLVM, or disabled explicitly by user.
* `WITH_XXX_BACKEND` affects whether the runtime backend is built
* `llvm-to-backend`-backend-specific infrastructure for the SSCP compiler is built if `WITH_XXX_BACKEND` is enabled.

Reply:
I want to signify that the Metal backend doesn't work without SSCP. For example, CMake should handle the following combos as:
- `WITH_SSCP_COMPILER` and `WITH_METAL_BACKEND` - builds GPU backend on ARM, error on x86
- `WITH_SSCP_COMPILER` - builds CPU-only everywhere
- `WITH_METAL_BACKEND` - error because Metal requires SSCP

For the first, I'd like to set `WITH_METAL_BACKEND` to true behind-the-scenes. It would only be true on ARM, but it would force Metal to be accessible on those platforms. I fear that developers will explicitly enable ze/cuda/rocm in their developer workflows, but forget Metal because macOS isn't their priority (they don't even know Metal exists). Therefore Metal will have 2nd-class support. Also, the goal of SSCP is reducing build complexity and the need to think about separate backends.

Reply:
So, metal is not unique with regards to only being able to run kernels from the SSCP compiler. The same is kind of true for Intel GPUs and the Level Zero backend (technically it can also still be targeted using the clang SYCL toolchain, but that functionality is very limited, is already behind the SSCP compiler in terms of support and will be replaced eventually).

I think there are use cases where runtime backend capabilities can be needed without the ability to compile user kernels. For example, a user might want to run SYCL kernels on the CPU exclusively, but might want to interop with some other Metal code, and would like to synchronize those operations with SYCL operations. This can be achieved using backend interoperability in SYCL, if the runtime has a Metal backend.

I think your use cases can be addressed using sensible defaults and auto-detection of system capabilities, can't they?

I'm not sure I understand your concerns correctly, but if you are concerned that many people might not be using/enabling the SSCP compiler, I can assure you that it is a crucial component and focus for the future of the project. Once it is mature enough, the idea is to position it as the default compilation flow and let it take center stage. Because of its unified code representation, it wouldn't matter if people ultimately run on metal or some other backends as long as they use the SSCP compiler.

Reply:
> but if you are concerned that many people might not be using/enabling the SSCP compiler

My concern is that many people use the SSCP compiler (no one's going to use the old workflow), but few people use the SSCP compiler + the Metal backend. Macs would be stuck in CPU-only because Metal is disabled by default. That's already the case with CUDA/ROCM/ZE workflows - you must explicitly *enable* each backend. I don't want Metal to be like that.

For example, some enters the hipSYCL directory and types `cmake .. -DWITH_SSCP_COMPILER -DWITH_HIP_BACKEND`. Could you provide a breakdown of what happens with each OS and with each hardware vendor?

Reply:
> My concern is that many people use the SSCP compiler (no one's going to use the old workflow), but few people use the SSCP compiler + the Metal backend. Macs would be stuck in CPU-only because Metal is disabled by default. That's already the case with CUDA/ROCM/ZE workflows - you must explicitly enable each backend. I don't want Metal to be like that.

This is absolutely not true. Those backends will automatically be enabled whenever CUDA or ROCm installations are found by cmake (this is not yet the case for Level Zero, but that is only because it was super experimental with the old compilation flow anyway).
This is pretty much the best we can do because we *have* to rely on preinstalled CUDA or ROCm anyway because those dependencies are too complex to ship with hipSYCL. If cmake cannot find it, we can only assume it's not installed or installed in some non-standard path, in which case the user must provide explicit cmake options anyway.

If shipping Metal dependencies with hipSYCL is simpler, or we can just assume it to be available on Mac, I am not opposed to just enabling it unconditionally for those platforms.

> For example, some enters the hipSYCL directory and types cmake .. -DWITH_SSCP_COMPILER -DWITH_HIP_BACKEND. Could you provide a breakdown of what happens with each OS and with each hardware vendor?

* `-DWITH_SSCP_COMPILER` (enabled by default whenever cmake finds an LLVM >= 14 on the system): 
  * If no sufficiently new LLVM is found it errors, otherwise it builds the core SSCP compiler infrastructure
* `-DWITH_ROCM_BACKEND` (enabled by default whenever cmake finds ROCm):
  * If no ROCm is found it errors, otherwise builds `librt-backend-hip` and `llvm-to-amdgpu` since `WITH_SSCP_COMPILER` is set.
 
I don't think there are any particular OS-specific branches regarding the HIP backend or SSCP compiler, but of course AMD supports ROCm only on Linux currently, so it will not be found or be installable on other OSes. But that is an AMD problem, not a hipSYCL problem ;)